sound/soc/sof/topology.c

```diff
@@ -2149,10 +2158,15 @@ static int sof_link_hda_process(struct snd_sof_dev *sdev,
 		if (strcmp(link->name, dai->name) == 0) {
 			config->dai_index = dai->comp_dai.dai_index;
 
-			if (dai->comp_dai.direction == 0)
+			if (dai->comp_dai.direction == 0) {
 				config->hda.link_dma_ch = tx_slot;
-			else
+				ret = sof_set_dai_config(sdev, size, link,
+							 config, NULL);
+			} else {
 				config->hda.link_dma_ch = rx_slot;
+				ret = sof_set_dai_config(sdev, size, link,
+							 NULL, config);
+			}
```

Comment:
What I meant is try to pass link and config ONCE, then use it internablly in sof_set_dai_config.
Passing two arguments, once of which is NULL with additional tests is just horrible and makes the code more confusing.

Reply:
for HDA, two different configs need to be set by sof_set_dai_config. 
If I just pass config ONCE, I need to add direction and call sof_set_dai_config two times.
The direction info is not included by link setting or config.

I will try to refine it without NULL.